Fix, update and prettify cURL information: `phpinfo()` and `curl_version()`

nono303 opened the pull request:
- Remove empty, null & deprecated values  
- Ensure displaying human readable values   
- Fix typo (like **ares** version was displayed as **ZLib**)  
- Fix non-working condition with `CURLVERSION_NOW` using `age`  
- Use `feature_names` if available instead of `const feats[]` and display features as list (like for protocols)  
- Update (legacy) duplicated `const feats[]` with `CURL_VERSION_CURLDEBUG`, `CURL_VERSION_THREADSAFE`  
- Display SSL supported backends  
- Add new supported protocols (**Brotli**, **Nghttp2**, **QUIC**, **Zstd**, **Hyper**, **GNU SASL**, **RTMP**)  
- Return protocols module versions in an array for `curl_version()`  
> Result with same **cURL 8.12.1** and **PHP 8.4.4**:  
### `phpinfo()`  
#### before  
![phpinfo_before](https://github.com/user-attachments/assets/02d8fd88-990a-4a0c-a8bf-59f187184947)  
#### after  
![phpinfo_after](https://github.com/user-attachments/assets/975f62fd-0d75-4165-8acc-02ed720c5bb2)  
### `curl_version()`  
#### before  
``````  
[version_number] => 527361  
[age] => 11  
[features] => 1575833501  
[feature_list] => Array  
(  
    [AsynchDNS] => 1  
    [CharConv] =>  
    [Debug] =>  
    [GSS-Negotiate] =>  
    [IDN] => 1  
    [IPv6] => 1  
    [krb4] =>  
    [Largefile] => 1  
    [libz] => 1  
    [NTLM] => 1  
    [NTLMWB] =>  
    [SPNEGO] => 1  
    [SSL] => 1  
    [SSPI] => 1  
    [TLS-SRP] => 1  
    [HTTP2] => 1  
    [GSSAPI] =>  
    [KERBEROS5] => 1  
    [UNIX_SOCKETS] => 1  
    [PSL] =>  
    [HTTPS_PROXY] => 1  
    [MULTI_SSL] => 1  
    [BROTLI] => 1  
    [ALTSVC] => 1  
    [HTTP3] =>  
    [UNICODE] => 1  
    [ZSTD] => 1  
    [HSTS] => 1  
    [GSASL] =>  
)  
  
[ssl_version_number] => 0  
[version] => 8.12.1-DEV  
[host] => Windows  
[ssl_version] => (OpenSSL/3.4.1) Schannel  
[libz_version] => 1.3.1  
[protocols] => Array  
(  
    [0] => dict  
    [1] => file  
    [2] => ftp  
    [3] => ftps  
    [4] => gopher  
    [5] => gophers  
    [6] => http  
    [7] => https  
    [8] => imap  
    [9] => imaps  
    [10] => ldap  
    [11] => ldaps  
    [12] => mqtt  
    [13] => pop3  
    [14] => pop3s  
    [15] => rtsp  
    [16] => scp  
    [17] => sftp  
    [18] => smb  
    [19] => smbs  
    [20] => smtp  
    [21] => smtps  
    [22] => telnet  
    [23] => tftp  
    [24] => ws  
    [25] => wss  
)  
  
[ares] => 1.34.4  
[ares_num] => 74244  
[libidn] =>  
[iconv_ver_num] => 0  
[libssh_version] => libssh2/1.11.1_DEV  
[brotli_ver_num] => 16781312  
[brotli_version] => brotli/1.1.0  
``````  
  
#### after  
  
``````  
[version] => 8.12.1-DEV  
[version_number] => 527361  
[age] => 11  
[host] => Windows  
[features] => 1575833501  
[feature_list] => Array  
(  
    [0] => alt-svc  
    [1] => AsynchDNS  
    [2] => brotli  
    [3] => HSTS  
    [4] => HTTP2  
    [5] => HTTPS-proxy  
    [6] => HTTPSRR  
    [7] => IDN  
    [8] => IPv6  
    [9] => Kerberos  
    [10] => Largefile  
    [11] => libz  
    [12] => MultiSSL  
    [13] => NTLM  
    [14] => SPNEGO  
    [15] => SSL  
    [16] => SSLS-EXPORT  
    [17] => SSPI  
    [18] => threadsafe  
    [19] => TLS-SRP  
    [20] => Unicode  
    [21] => UnixSockets  
    [22] => zstd  
)  
  
[protocols] => Array  
(  
    [0] => dict  
    [1] => file  
    [2] => ftp  
    [3] => ftps  
    [4] => gopher  
    [5] => gophers  
    [6] => http  
    [7] => https  
    [8] => imap  
    [9] => imaps  
    [10] => ldap  
    [11] => ldaps  
    [12] => mqtt  
    [13] => pop3  
    [14] => pop3s  
    [15] => rtsp  
    [16] => scp  
    [17] => sftp  
    [18] => smb  
    [19] => smbs  
    [20] => smtp  
    [21] => smtps  
    [22] => telnet  
    [23] => tftp  
    [24] => ws  
    [25] => wss  
)  
  
[ssl_backends] => Array  
(  
    [0] => openssl  
    [1] => schannel  
)  
  
[feature_version] => Array  
(  
    [ssl] => OpenSSL/3.4.1 (Schannel)  
    [libz] => 1.3.1  
    [ares] => 1.34.4  
    [libssh] => libssh2/1.11.1_DEV  
    [brotli] => brotli/1.1.0  
    [nghttp2] => 1.64.0  
    [zstd] => zstd/1.5.6  
)  
``````

Commit:
Update duplicated `const feats[]` with `CURL_VERSION_CURLDEBUG`, `CURL_VERSION_THREADSAFE`

Commit:
Use `feature_names` if available instead of `const feats[]` and display features as list (like for protocols)

Commit:
Display SSL supported backends

Commit:
Fix typo (like **ares** version was displayed as **ZLib**)

Commit:
remove non human readable values (might be discussed)

Commit:
Fix non-working condition with `CURLVERSION_NOW` using `age`

Commit:
fix feature_names

Commit:
Return protocols module versions in an array for `curl_version()`

Commit:
Add new supported protocols (**Brotli**, **Nghttp2**, **QUIC**, **Zstd**, **Hyper**, **GNU SASL**, **RTMP**)

Commit:
clean indent: space > tab

Commit:
rollback for curl_version()
- put back long: version_number & features
- keep original key: feature_list

cmb69 commented:
Thank you for the PR! I'm generally in favor of these changes, but we need to cater to BC, so ideally all tests should pass without needing to be modified. It might also make sense to explicitly check the changes against libcurl 7.61.0 which is the oldest version we're supporting (I presume that none of the CI jobs runs against such an old version).

Minor issue: it might be reasonable to not introduce unrelated whitespace changes (these can be done in a separate PR, if at all).

cmb69 commented:
cc @Ayesh

nono303 commented:
Thx @cmb69 for your feedback!
> ll tests should pass without needing to be modified

I'll have a look on what failed

> It might also make sense to explicitly check the changes against libcurl 7.61.0

Hmmmm... actually and out of my PR with php 8.4.4, libcurl 7.61.0 doesn't work 
 - `curl_easy_upkeep` required by ext/curl are is not present _(at least exposed by shared libcurl)_ so it'll be difficult for me to do this check
![](https://github.com/user-attachments/assets/2d14cd07-f498-45a2-a7e4-b9206fd7b2b8)
```S:\telecharger>curl --version
curl 7.61.0-DEV (Windows) libcurl/7.61.0-DEV OpenSSL/3.4.0a (WinSSL) zlib/1.3.1 brotli/1.1.0 c-ares/1.34.4 libssh2/1.11.1_DEV nghttp2/1.64.0
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher http https imap imaps ldap pop3 pop3s rtsp scp sftp smtp smtps telnet tftp
Features: AsynchDNS IPv6 Largefile SSPI Kerberos SPNEGO NTLM SSL libz brotli HTTP2 HTTPS-proxy MultiSSL
```
> not introduce unrelated whitespace changes

I does it for my change (only space after \n...) but as I saw some other few inconsistencies, I've just fixed them afterwards... but I can revert it if needed

cmb69 commented:
> I'll have a look on what failed

Thanks!

> `curl_easy_upkeep` required by ext/curl are is not present _(at least exposed by shared libcurl)_ so it'll be difficult for me to do this check

The code using `curl_easy_upkeep()` is not supposed to be compiled when built against libcurl < 7.62.0. Anyway, it's not strictly necessary that *you* do this check; it's just that somebody better should double-check it (I don't think we can bump the cURL requirement further for now). If need be, I think I can do that.

> I does it for my change (only space after \n...) but as I saw some other few inconsistencies, I've just fixed them afterwards... but I can revert it if needed

I, personally, would prefer if you revert the unrelated whitespace changes (keeps the diff a bit smaller, and more focused). But maybe wait a while what others think about that.

nono303 commented:
... I just temporary commented part of `curl_easy_upkeep()` to check against 7.61 ;)

Looking to the failing test, I would have to rollback **[Fix non-working condition with CURLVERSION_NOW using age](https://github.com/php/php-src/pull/17848/commits/44ec7e7d23107e1a1ef0c4941b26ddf170628e67)** `curl_version_info_data->age` to `CURLVERSION_XXX` testing but as mentionned ,their is an issue:
only `CURLVERSION_NOW` is defined. All other (`CURLVERSION_FIRST`...) are well declared in curl.h `CURLversion enum` but not defined.

🔴  This is why all test `#if defined(CURLVERSION_XXX) && CURLVERSION_NOW >= CURLVERSION_XXX` never worked and module version were not displayed in both `phpinfo()` and `curl_version()`

```c
#if defined(CURLVERSION_NOW)
	#pragma message ("*** CURLVERSION_NOW defined")
#else
	#pragma message ("*** CURLVERSION_NOW NOT defined")
#endif
#if defined(CURLVERSION_SECOND)
	#pragma message ("*** CURLVERSION_SECOND defined")
#else
	#pragma message ("*** CURLVERSION_SECOND NOT defined")
#endif
```
```
2025-02-19_09-26-51 ####### BEGIN BUILD MODULE 'php_curl' vs17 x64 avx2 ###########################
curl_file.c
interface.c
multi.c
share.c
*** CURLVERSION_NOW defined
*** CURLVERSION_SECOND NOT defined
   Creating library C:\Program Files\php\php_curl.lib and object C:\Program Files\php\php_curl.exp
Generating code
Finished generating code
```
Does I missed something??

Commit:
use LIBCURL_VERSION_NUM to fix https://github.com/php/php-src/pull/17848/commits/44ec7e7d23107e1a1ef0c4941b26ddf170628e67

Commit:
fix incompatible integer to pointer conversion passing 'int' to parameter of type 'const char *'

Commit:
typo

Commit:
BC:
 - remove ssl_version_number: not used anymore, always 0 https://github.com/curl/curl/blob/master/include/curl/curl.h#L3089
 - ["ssl_version"] & ["libz_version"] moved to ["features_version"]["ssl"] & ["features_version"]["libz"]

Commit:
typo

cmb69 commented on ext/curl/interface.c:
```diff
@@ -1028,104 +1088,173 @@ PHP_FUNCTION(curl_version)
-				add_assoc_bool(&feature_list, feats[i].name, d->features & feats[i].bitmask ? true : false);
+	zval feature_list;
+	array_init(&feature_list);
+#if defined(CURLVERSION_ELEVENTH) && CURLVERSION_NOW >= CURLVERSION_ELEVENTH
```
`CURLVERSION_ELEVENTH` etc. are enum values (https://github.com/curl/curl/blob/ee3f657607e32bf6c252e1732b47405be065881f/include/curl/curl.h#L3059-L3073), and as such can't be used in preprocessor conditionals. Instead use the numbers (`CURLVERSION_ELEVENTH` is `0x075700` etc.)

nono303 replied:
This is what I noticed and didn’t understand why one part was managed with `CURLVERSION_*` (phpinfo() curl_version()) and another with `LIBCURL_VERSION_NUM` (struct feat feats[])... 
✅ fixed & working now for 7.61 & 8.12

❓ Would adding this to `curl_private.h` would make sense?
```
#define PHP_CURLVERSION_TWELFTH=	LIBCURL_VERSION_NUM >= 0x080800 /* Available since 8.8.0	*/	
#define PHP_CURLVERSION_ELEVENTH =	LIBCURL_VERSION_NUM >= 0x075700 /* Available since 7.87.0	*/	
#define PHP_CURLVERSION_TENTH =		LIBCURL_VERSION_NUM >= 0x074d00 /* Available since 7.77.0	*/	
#define PHP_CURLVERSION_NINTH =		LIBCURL_VERSION_NUM >= 0x074b00 /* Available since 7.75.0	*/	
#define PHP_CURLVERSION_EIGHTH =	LIBCURL_VERSION_NUM >= 0x074800 /* Available since 7.72.0	*/	
#define PHP_CURLVERSION_SEVENTH  =	LIBCURL_VERSION_NUM >= 0x074600 /* Available since 7.71.0	*/	
#define PHP_CURLVERSION_SIXTH =		LIBCURL_VERSION_NUM >= 0x074200 /* Available since 7.66.0	*/	
#define PHP_CURLVERSION_FIFTH =		LIBCURL_VERSION_NUM >= 0x073900 /* Available since 7.57.0	*/	
#define PHP_CURLVERSION_FOURTH =	LIBCURL_VERSION_NUM >= 0x071001 /* Available since 7.16.1	*/	
#define PHP_CURLVERSION_THIRD =		LIBCURL_VERSION_NUM >= 0x070C00 /* Available since 7.12.0	*/	
#define PHP_CURLVERSION_SECOND =	LIBCURL_VERSION_NUM >= 0x070B01 /* Available since 7.11.1	*/	
#define PHP_CURLVERSION_FIRST =		LIBCURL_VERSION_NUM >= 0x070A00 /* Available since 7.10		*/
```
Besides, I continue the my adjustments and will keep you informed when it would be done for me.

cmb69 replied:
`CURLVERSION_*` especially refers to `struct curl_version_info_data`, while `LIBCURL_VERSION_NUM` is the actual cURL version (the former is only increased if the struct declaration changes).

Commit:
fix curl/tests/curl_version_features-array.phpt with new feature_list

nono303 commented:
❓ Would adding this to `curl_private.h` would make sense? _(cf. https://github.com/php/php-src/pull/17848#discussion_r1963145831)_
```
#define PHP_CURLVERSION_TWELFTH=	LIBCURL_VERSION_NUM >= 0x080800 /* Available since 8.8.0	*/	
#define PHP_CURLVERSION_ELEVENTH =	LIBCURL_VERSION_NUM >= 0x075700 /* Available since 7.87.0	*/	
#define PHP_CURLVERSION_TENTH =		LIBCURL_VERSION_NUM >= 0x074d00 /* Available since 7.77.0	*/	
#define PHP_CURLVERSION_NINTH =		LIBCURL_VERSION_NUM >= 0x074b00 /* Available since 7.75.0	*/	
#define PHP_CURLVERSION_EIGHTH =	LIBCURL_VERSION_NUM >= 0x074800 /* Available since 7.72.0	*/	
#define PHP_CURLVERSION_SEVENTH  =	LIBCURL_VERSION_NUM >= 0x074600 /* Available since 7.71.0	*/	
#define PHP_CURLVERSION_SIXTH =		LIBCURL_VERSION_NUM >= 0x074200 /* Available since 7.66.0	*/	
#define PHP_CURLVERSION_FIFTH =		LIBCURL_VERSION_NUM >= 0x073900 /* Available since 7.57.0	*/	
#define PHP_CURLVERSION_FOURTH =	LIBCURL_VERSION_NUM >= 0x071001 /* Available since 7.16.1	*/	
#define PHP_CURLVERSION_THIRD =		LIBCURL_VERSION_NUM >= 0x070C00 /* Available since 7.12.0	*/	
#define PHP_CURLVERSION_SECOND =	LIBCURL_VERSION_NUM >= 0x070B01 /* Available since 7.11.1	*/	
#define PHP_CURLVERSION_FIRST =		LIBCURL_VERSION_NUM >= 0x070A00 /* Available since 7.10		*/
```

Commit:
fix check_win_config.phpt

cmb69 commented:
> Would adding this to `curl_private.h` make sense

While I don't find the hexadecimal numeric cURL version number not super readable, `CURLVERSION_*` is incomprehensable. I would stick with the hard-coded numbers.

Commit:
fix last usage of CURLVERSION_*

Commit:
fix curl_version_features-array for  LIBCURL_VERSION_NUM >= 7.87.0 (only enabled features are listed)

Commit:
update expected based on LINUX_X64_RELEASE_NTS (as curl implementation may / are different between OS)
